### PR TITLE
Cleanup inline_const

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

File: src/runtime/locations.jl [hunk 1]
```diff
@@ -116,15 +116,27 @@ function map_check_nothrow(parent::Locations{NTuple{N,Int}}, sub::Locations{Unit
...
+function map_check_nothrow(parent::Locations{UnitRange{Int}}, sub::Locations{NTuple{N,Int}}) where {N}
     all(x -> (1 <= x <= length(parent)), sub.storage)
 end
 
-function map_check(parent::Locations{UnitRange{Int}}, sub::Locations{UnitRange{Int}})
+@noinline function map_check(parent::Locations{UnitRange{Int}}, sub::Locations{NTuple{N,Int}}) where {N}
+    all(x -> (1 <= x <= length(parent)), sub.storage)
+end
+
+function map_check_throw(parent::Locations{UnitRange{Int}}, sub::Locations{UnitRange{Int}})
```
Comment (vchuravy) [inline src/runtime/locations.jl]:
```suggestion
function map_check_nothrow(parent::Locations{UnitRange{Int}}, sub::Locations{UnitRange{Int}})
```

**Comment by Roger-luo** [comment]
OK I think I know why the codegen doesn't work, this is because the codegen doesn't know how to handle branches other than `if <var> = <const>`, after the `inline_const!` and `elim_map_check!` pass, we get a bunch of trivial branches like the following

```julia
80 ──        invoke YaoCompiler.Semantic.gate(HGate::YaoCompiler.Intrinsics.HGate, $(QuoteNode(Locations(3)))::Locations{Int64})::Nothing
   └────        goto #81                                            │││╻╷╷╷╷   unsafe_mapping
   81 ──        goto #82                                            ││││┃│││    map
   82 ──        goto #83                                            │││││┃│      #3
   83 ──        goto #84                                            ││││││  
   84 ──        goto #85                                            ││││││╻╷      getindex
   85 ──        goto #86                                            │││││││┃       getindex
   86 ──        goto #87                                            │││││││ 
   87 ──        goto #88                                            ││││││  
   88 ──        goto #89                                            │││││   
   89 ──        goto #90                                            ││││    
   90 ──        goto #91      
```

I tried to run `compact!(ir, true)` again on this, but it seems it won't simplify this CFG?

**Comment by vchuravy** [comment]
https://github.com/QuantumBFS/YaoCompiler.jl/pull/9/commits/edba84c12c465ac507abca7d7292c929dc5007f9#diff-073e1737aab44dd56bb73b786134a351240aa2ecaf838462d71a32020f1640c7R105

should skip those

**Comment by Roger-luo** [comment]
Yay! works now

```julia
julia> ast = @code_qasm optimize = true circuit()
OPENQASM 2.0
qreg qreg_1[3];
h qreg_1[2];
CX qreg_1[1], qreg_1[2];
tdg qreg_1[2];
CX qreg_1[0], qreg_1[2];
t qreg_1[2];
CX qreg_1[1], qreg_1[2];
tdg qreg_1[2];
CX qreg_1[0], qreg_1[2];
t qreg_1[1];
t qreg_1[2];
h qreg_1[2];
CX qreg_1[0], qreg_1[1];
t qreg_1[0];
tdg qreg_1[1];
CX qreg_1[0], qreg_1[1];
```

**Comment by Roger-luo** [comment]
I just fixed the codegen, I wrote a wrong index for `measure` statement in typed CodeInfo, now this PR should work. But there still seems to be some issues in the `group_quantum_stmt!` pass, `group_quantum_stmts!(ir)`
this will gives an out of bounds error if enabled, I guess this is because I didn't implement the `permute_stmts` correctly? https://github.com/QuantumBFS/YaoCompiler.jl/blob/master/src/compiler/optimize.jl#L86

**Pull request merged**